File: Cracking the Coding Interview 6th Edition/Chapter 1/URLify/URLify.cpp
#include <string>
#include <iostream>

using namespace std;

void URLify(string& input, int trueLength) {
	int back = input.size() - 1;

	for (int i = trueLength - 1; i >= 0; --i) {
		char& cur = input.at(i);

		if (cur == ' ') {
			input.at(back--) = '0';
			input.at(back--) = '2';
			input.at(back--) = '%';
		} else {
			input.at(back--) = cur;
		}
	}
}

int main() {
	string test = "abc d  ";
	URLify(test, 5);
	cout << test << endl;
}

File: Cracking the Coding Interview 6th Edition/Chapter 1: Arrays and Strings/URLify/URLify.py
def URLify(s, n):
	lci = n - 1

	for i in range(n - 1, 0, -1):
		if s[i] != ' ':
			lci = i
			break

	back = len(s) - 1
	for j in range(lci, 0, -1):
		if s[j] == ' ':
			s[back - 2 : back + 1] = '%20'
			back -= 3
		else:
			s[back] = s[j]
			back -= 1

s = list('a b c    ') #'a%20b%20c'
n = 9
URLify(s, n)
print(s)